File: Blind 75/13 - 1D Dynamic Programming/198-house-robber.py
'''
You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
'''
from time import time
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        # There are two choices:
        # 1) Rob nums[i] + nums[i - 2] (current and before previous)
        # 2) Rob nums[i - 1] (previous)
        # The max of these two choices is the max profit
        prev, curr = 0, 0
        for n in nums:
            # After optimizing for O(1) space
            prev, curr = curr, max(curr, prev + n)

        return curr

    def reference(self, nums: List[int]) -> int:
        plan_a, plan_b = 0, 0

        for n in nums:
            temp = max(n + plan_a, plan_b)
            plan_a = plan_b
            plan_b = temp

        return plan_b

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.rob(case))
                else:
                    self.rob(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [[1, 2, 3, 1], [2, 7, 9, 3, 1]]
    test.quantify(test_cases)

File: Blind 75/13 - 1D Dynamic Programming/JavaScript/198-house-robber.js
const { performance } = require('perf_hooks');

class Solution {
  method(nums) {
    let [prev, curr] = [0, 0];
    for (const n of nums) {
      [prev, curr] = [curr, Math.max(curr, prev + n)];
    }

    return curr;
  }

  reference(nums, i = 0, memo = this.initMemo(nums)) {
    const isBaseCase = nums.length <= i;
    if (isBaseCase) return 0;

    const hasSeen = 0 <= memo[i];
    if (hasSeen) return memo[i];

    const [next, nextNext] = [i + 1, i + 2];
    const right = nums[i];
    const mid = this.reference(nums, next, memo); /* Time O(N) | Space O(N) */
    const left = this.reference(
      nums,
      nextNext,
      memo
    ); /* Time O(N) | Space O(N) */
    const house = left + right;

    memo[i] = Math.max(mid, house); /*           | Space O(N) */

    return memo[i];
  }

  initMemo = (nums) => Array(nums.length + 1).fill(-1);

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [1, 2, 3, 1],
  [2, 7, 9, 3, 1],
];
test.quantify(testCases);
